### DISCUSSION TITLE
How to set a slave ID for ModbusSimulator context?

### DISCUSSION THREAD

**Discussion by schwarzkunstlerin** [pymodbus-dev/pymodbus#2040] [Help] [answered]
Greetings!
I am not sure how to do that or if it is possible to be done, so I need more clarification on that topic.

Many real-life Modbus devices need a slave ID to be connected with. That is why we need `slave` when we try to read some registers from the device -> `client.read_holding_registers(address, count, slave)`.

When I try to build Modbus Simulator with JSON & ModbusSimulatorContext, I do not provide in the config any slave ID under which this simulator will be available - in this case, every client's request done on any address will be successful, because the simulator works on broadcast mode & every slave ID returns the same context.

I have read from the docs that it is possible to change that in ModbusServerContext - I can set `single` parameter to `False` and then it will be interpeted as a collection of slaves context that are under the specific slave ID, so it will be possible to set a specific slave ID for my simulator? 

The point is, I have the task of writing a simulator of a real device, and this real device always works under a specific slave ID. I expect that in this case the client will receive a response from the server-simulator only when I provide the correct slave ID in the client and not for any random slave.

I have noticed that ModbusServerContext has a method `slaves()` to define slaves (?), but how can I use that? Is it possible to define a slave ID for ModbusSimulator? Any tips on how to achieve the results mentioned above in my message are also welcome.

My current code is simple, based on ModbusSimulator examples:
````
config = config = {
    "setup": {
        "co size": 100,
        "di size": 100,
        "hr size": 2000,
        "ir size": 2000,
        "shared blocks": True,
        "type exception": False,
        "defaults": {
            "value": {"bits": 0, "uint16": 0, "uint32": 0, "float32": 0.0, "string": " "},
            "action": {"bits": None, "uint16": None, "uint32": None, "float32": None, "string": None},
        },
    },
    "invalid": [],
    "write": [[8, 8]],
    "bits": [],
    "uint16": [
        {"addr": [8, 8], "value": 10},
    ],
    "uint32": [
        {"addr": [96, 97], "value": 999},
    ],
    "float32": [],
    "string": [],
    "repeat": [],
}

context = ModbusSimulatorContext(config=config, custom_actions={})
context = ModbusServerContext(slaves=context, single=False)  # what now?
...
StartAsyncTcpServer(context=context, address=(host, port), framer=framer)
```

**Comment by janiversen** [comment]
Simulator datastore simulates 1 slave therefore the slave id is not needed.

Unless you want to simulate a multidrop line, you can just use that context. In all other cases it does not produce problems that the simulator responds to to any id.

You can probably combine ModbusServerContext and ModbusSimulatorStore to create a multidrop line, but I have never tried it.

**Comment by janiversen** [comment]
You can see in our examples how to make a server context with multiple slaves (multiple can also be just one).

**Comment by janiversen** [comment, marked as answer]
Just checked.

You can define 1 or multiple ModbusSimulatorContext, and assign them with slave ids to the server.

There are several examples of how to assign multiple contexts with slave ids to a server, have a look.

**Reply by schwarzkunstlerin** [reply]
Thank you very much for your help, I managed to achieve the expected behavior.
